torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,111 @@
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD gradient compression
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution and orthogonalized;
+    2) Allreduces P;
+    3) Orthogonizes P;
+    4) Compute Q, which is approximately equal to M^TP;
+    5) Allreduces Q;
+    6) Computes M, which is approximately equal to PQ^T.
+    7) Truncates the input tensor to the original length.
+
+    TODO(wayi@): The above procedure does two matmul+allreduce steps per iteration --
+    one left multiplication and one right multiplication.
+    For warm start, can take one such step at a time, and alternate between them.
+
+    Example::
+        PowerSGDState state(process_group, 1)
+        >>> ddp_model.register_comm_hook(state, powerSGD_hook)
+    """
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+    world_size = (
+        process_group.size() if process_group is not None else dist.get_world_size()
+    )
+
+    # The input tensor is a flattened 1D tensor.
+    input_tensor = bucket.get_tensors()[0]
+    device = input_tensor.device
+    total_length = input_tensor.shape[0]
+
+    # View the input tensor as a 2D square-shape tensor, and pad 0s if necessary.
+    square_side_length = math.ceil(math.sqrt(total_length))
+    padded_total_length = square_side_length ** 2
+    input_tensor.resize_(padded_total_length)
+    input_tensor[total_length:padded_total_length].fill_(0)
+    matrix = input_tensor.view(square_side_length, square_side_length)
+
+    def create_low_rank_tensor(fill_random_values):
+        "Returns a low-rank 2D tensor and the allocated contiguous memory."
+        memory = torch.empty(square_side_length * rank, device=device)
+        if fill_random_values:
+            with torch.random.fork_rng():
+                # The seed makes sure that the initial random values are the same across all the DDP replicas.
+                # Such seed should differ at every step.
+                # Currently use the length of input tensor as the seed, which should be mostly different.
+                # TODO(wayi@): Should read the random seed from the state of this hook provided by the constructor.
+                torch.manual_seed(total_length)
+                memory[:] = torch.randn_like(memory)
+        return memory.view(square_side_length, rank), memory
```

Comment:
Should `rank` be an input into this function rather than implicitly captured from the context? Also, should we have unittests for this function?